blockindex/sgd_indexer.go

```diff
@@ -23,34 +194,238 @@ type (
+	}
+}
+
 // NewSGDRegistry creates a new SGDIndexer
-func NewSGDRegistry() SGDRegistry {
-	return &sgdRegistry{}
+func NewSGDRegistry(contract string, kv db.KVStore) SGDRegistry {
+	if kv == nil {
+		panic("nil kvstore")
+	}
+	return &sgdRegistry{
+		contract: contract,
+		kvStore:  kv,
+	}
 }
 
+// Start starts the SGDIndexer
 func (sgd *sgdRegistry) Start(ctx context.Context) error {
-	return nil
+	return sgd.kvStore.Start(ctx)
 }
 
+// Stop stops the SGDIndexer
 func (sgd *sgdRegistry) Stop(ctx context.Context) error {
-	return nil
+	return sgd.kvStore.Stop(ctx)
 }
 
+// Height returns the current height of the SGDIndexer
 func (sgd *sgdRegistry) Height() (uint64, error) {
-	return 0, nil
+	h, err := sgd.kvStore.Get(_sgdToHeightNS, _sgdCurrentHeight)
+	if err != nil {
+		// if db not exist, return 0, nil, after PutBlock, the height will be increased
+		if errors.Is(err, db.ErrNotExist) {
+			return 0, nil
+		}
+		return 0, err
+	}
+	return byteutil.BytesToUint64BigEndian(h), nil
 }
 
+// PutBlock puts a block into SGDIndexer
 func (sgd *sgdRegistry) PutBlock(ctx context.Context, blk *block.Block) error {
+	var (
+		r  *action.Receipt
+		ok bool
+	)
+	b := batch.NewBatch()
+	receipts := getReceiptsFromBlock(blk)
+	for _, selp := range blk.Actions {
+		actHash, err := selp.Hash()
+		if err != nil {
+			continue
+		}
+		r, ok = receipts[actHash]
+		if !ok || r.Status != uint64(iotextypes.ReceiptStatus_Success) {
+			continue
+		}
+		for _, log := range r.Logs() {
+			if log.Address != sgd.contract {
+				continue
+			}
+			if err := sgd.handleEvent(b, log); err != nil {
+				return err
+			}
+		}
+	}
+	b.Put(_sgdToHeightNS, _sgdCurrentHeight, byteutil.Uint64ToBytesBigEndian(blk.Height()), "failed to put current height")
+	return sgd.kvStore.WriteBatch(b)
+}
+
+func (sgd *sgdRegistry) handleEvent(b batch.KVStoreBatch, log *action.Log) error {
+	abiEvent, err := _sgdABI.EventByID(common.Hash(log.Topics[0]))
+	if err != nil {
+		return err
+	}
+	switch abiEvent.Name {
+	case "ContractRegistered":
+		return sgd.handleContractRegistered(b, log)
+	case "ContractApproved":
+		return sgd.handleContractApproved(b, log)
+	case "ContractUnapproved":
+		return sgd.handleContractUnapproved(b, log)
+	case "ContractRemoved":
+		return sgd.handleContractRemoved(b, log)
+	default:
+		//skip other events
+	}
 	return nil
 }
 
-func (sgd *sgdRegistry) DeleteTipBlock(context.Context, *block.Block) error {
+func (sgd *sgdRegistry) handleContractRegistered(b batch.KVStoreBatch, log *action.Log) error {
+	var (
+		sgdIndex *indexpb.SGDIndex
+		err      error
+		event    struct {
+			ContractAddress common.Address
+			Recipient       common.Address
+		}
+	)
+	if err := _sgdABI.UnpackIntoInterface(&event, "ContractRegistered", log.Data); err != nil {
+		return err
+	}
+	contract, err := address.FromBytes(event.ContractAddress.Bytes())
+	if err != nil {
+		return err
+	}
+	recipient, err := address.FromBytes(event.Recipient.Bytes())
+	if err != nil {
+		return err
+	}
```

Comment:
these 2 conversion not needed, by defining `contract` and `recipient` as `[]byte`